### ISSUE TITLE
Discriminator Loss Bug

### ISSUE THREAD

**Issue by kaihe** [CompVis/taming-transformers#137]
As refered in #93 , discriminator is the key to get sharp image. But in my experiments, sometime aeloss get negative values. 
![image](https://user-images.githubusercontent.com/7460776/151085947-a0e02ea1-6fbf-43ca-a70a-fac76bf163e6.png)
I think in any network design, negative loss should be avoid because it could go negative infinity. 

The negative term comes from g_loss as `g_loss = -torch.mean(logits_fake)` , where logits_fake is the value after convolution, no sigmoid to limit the outputs logit_values. When the generator successfully generate a correct patch such as pure white space, discriminator will find some way to exploit this patch by making its logit go infinity, which will encourage generator to draw more patches like this and completely ignore other loss terms.

After checking the original code in [cycle-gan repo](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/003efc4c8819de47ff11b5a0af7ba09aee7f5fc1/models/networks.py#L538),  
![image](https://user-images.githubusercontent.com/7460776/151087699-665564f1-017d-4d34-b308-b6c7a3103889.png)
There do exist an outer loss term to limit the patch logits, only if gan_mode in ['wgangp']. I did not see any fancy 'wgangp' thing, so this should be a bug.

**Comment by thuangb** [comment]
This is not a bug, the loss used in the paper is Hinge GAN loss, which can have negative values for g_loss. If trained stably and correctly, the loss won't go to negative infinity

**Comment by SuwoongHeo** [comment]
Hi, I`m curious about how the `g_loss` is derived without using a sigmoid or softplus function. After searching several hours, I cannot find any reference for using logits directly like in this implementation. 
https://github.com/CompVis/taming-transformers/blob/1bbc027acb6a47e4eb348d611f9af53f1038ffee/taming/modules/losses/vqperceptual.py#L98

**Comment by function2-llx** [comment]
> Hi, I`m curious about how the `g_loss` is derived without using a sigmoid or softplus function. After searching several hours, I cannot find any reference for using logits directly like in this implementation.
> 
> https://github.com/CompVis/taming-transformers/blob/1bbc027acb6a47e4eb348d611f9af53f1038ffee/taming/modules/losses/vqperceptual.py#L98

@SuwoongHeo That seems to be Wasserstein loss. I find an implementation [here](https://github.com/igul222/improved_wgan_training/blob/fa66c574a54c4916d27c55441d33753dcc78f6bc/gan_64x64.py#L482-L500) from the official repository of paper [Improved Training of Wasserstein GANs](https://arxiv.org/abs/1704.00028).

The description in the VQGAN paper seems inaccurate, which claims to use a binary cross entropy loss following [Patch GAN](https://github.com/znxlwm/pytorch-pix2pix/blob/3059f2af53324e77089bbcfc31279f01a38c40b8/pytorch_pix2pix.py#L138).

![image](https://github.com/CompVis/taming-transformers/assets/38486514/29ddac79-62a9-44c4-89f3-59447957147c)

**Comment by whchan05** [comment]
> > Hi, I`m curious about how the `g_loss` is derived without using a sigmoid or softplus function. After searching several hours, I cannot find any reference for using logits directly like in this implementation.
> > https://github.com/CompVis/taming-transformers/blob/1bbc027acb6a47e4eb348d611f9af53f1038ffee/taming/modules/losses/vqperceptual.py#L98
> 
> @SuwoongHeo That seems to be Wasserstein loss. I find an implementation [here](https://github.com/igul222/improved_wgan_training/blob/fa66c574a54c4916d27c55441d33753dcc78f6bc/gan_64x64.py#L482-L500) from the official repository of paper [Improved Training of Wasserstein GANs](https://arxiv.org/abs/1704.00028).
> 
> The description in the VQGAN paper seems inaccurate, which claims to use a binary cross entropy loss following [Patch GAN](https://github.com/znxlwm/pytorch-pix2pix/blob/3059f2af53324e77089bbcfc31279f01a38c40b8/pytorch_pix2pix.py#L138).
> 
> ![image](https://user-images.githubusercontent.com/38486514/250100371-29ddac79-62a9-44c4-89f3-59447957147c.png)

I wonder why they won't simply stick with what's said in the paper? I spent few hours trying to figure out why
I am lucky to stumble upon your comment but imagine if I started a few weeks earlier it would have been a nightmare
